src/linter/rule.zig

```diff
@@ -105,6 +112,176 @@ pub const Rule = struct {
             .ptr = ptr,
             .runOnNodeFn = gen.runOnNode,
             .runOnSymbolFn = gen.runOnSymbol,
+            .kind = .builtin,
+        };
+    }
+
+    fn hashName(name: []const u8) Rule.Id {
+        var hasher = std.hash.Wyhash.init(0);
+        std.hash.autoHashStrat(&hasher, name, .Deep);
+        const hash = hasher.final();
+        const repr = std.mem.bytesToValue(Rule.Id.Repr, std.mem.asBytes(&hash));
+        return Rule.Id.new(repr);
+    }
+
+    const UserDefinedData = struct {
+        dll_handle: *anyopaque,
+        maybe_runOnNode: ?*const fn (*const anyopaque, NodeWrapper, *LinterContext) u16,
+        maybe_runOnSymbol: ?*const fn (*const anyopaque, Symbol.Id, *LinterContext) u16,
+    };
+
+    pub fn initUserDefined(alloc: std.mem.Allocator, file_path: [:0]const u8) !Rule {
+        // NOTE: we should also provide a module so that people can define their own
+        // more complicated user-defined rule builds in their build.zig
+
+        // TODO: prompt the user when we have to build it for the first time
+        std.debug.print("compiling user defined plugin at '{s}'\n", .{file_path});
+
+        // NOTE: this will be slow because we block on it,
+        // we should be compiling several user defined rules in parallel
+        const user_entry_arg = try std.fmt.allocPrint(alloc, "-Muser_entry={s}", .{file_path});
+        defer alloc.free(user_entry_arg);
+
+        var args = std.process.args();
+        defer args.deinit();
+        const exe_path = args.next() orelse @panic("no exe arg on posix"); // FIXME: not cross platform, is there a better way anyway?
+
+        const abi_path = try std.fs.path.join(alloc, &.{ exe_path, "../../share/user-defined/custom_rule_api.zig" });
+        defer alloc.free(abi_path);
```

Comment:
Can we stick all these into an arena and just deinit once at the end?

Reply:
oh yes. This PR (especially this part) is not worth reviewing yet. I was just trying to get that part to work.

Really I think it would be better to output a module in your build.zig and instead require a path to the output or at least the build artifact name or something.